src/main/java/com/aphyr/riemann/client/MapPromise.java

```diff
@@ -85,6 +85,19 @@ public Object unsafeDeref(final long time,
   }
 
   @Override
+  public void setCallback(final Callback<T2> callback) {
+    this.p.setCallback(new Callback<T1>() {
+      public void call(T1 value, Exception e) {
+        if (e != null) {
+          callback.call(null, e);
+        }
+        Object mapped = mapCapturingExceptions(value);
```

Comment:
MapPromise is supposed to enforce that the map function is only called once. How about calling deref() here instead?